File: apps/worker/worker/documents/normalizer.py

import re
from collections.abc import Iterable, Mapping

SKILL_ALIASES: dict[str, tuple[str, ...]] = {
	"AWS": ("aws", "amazon web services"),
	"Docker": ("docker",),
	"JavaScript": ("javascript", "js"),
	"Kubernetes": ("kubernetes", "k8s"),
	"PostgreSQL": ("postgresql", "postgres"),
	"Python": ("python",),
	"React": ("react", "reactjs", "react.js"),
	"SQL": ("sql",),
	"TypeScript": ("typescript", "ts"),
}


def normalize_resume(blocks: Iterable[Mapping[str, object]]) -> dict[str, object]:
	block_list = list(blocks)
	return {
		"contact": contact_facts(block_list),
		"skills": [
			{"canonicalName": skill, "evidenceBlockIds": evidence}
			for skill, aliases in SKILL_ALIASES.items()
			if (evidence := matching_blocks(block_list, aliases))
		]
	}


def contact_facts(blocks: Iterable[Mapping[str, object]]) -> dict[str, str | None]:
	# Blocks are page- or document-granular; contact heuristics work on lines.
	lines = [
		line.strip()
		for block in blocks
		for line in str(block["text"]).splitlines()
		if line.strip()
	]
	combined = "\n".join(lines)
	email_match = re.search(r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b", combined, re.IGNORECASE)
	first_line = lines[0] if lines else ""
	name_words = first_line.split()
	name = (
		first_line
		if 2 <= len(name_words) <= 5
		and re.fullmatch(r"[A-Za-z][A-Za-z .'-]{1,80}", first_line)
		and all(word[0].isupper() for word in name_words)
		else None
	)
	location = next(
		(
			line.removeprefix("Location:").strip()
			for line in lines[:12]
			if line.casefold().startswith("location:")
		),
		None,
	)
	return {
		"name": name,
		"email": email_match.group(0) if email_match else None,
		"location": location,
	}


def matching_blocks(blocks: Iterable[Mapping[str, object]], aliases: Iterable[str]) -> list[str]:
	pattern = re.compile(
		"|".join(rf"(?<![a-z0-9]){re.escape(alias)}(?![a-z0-9])" for alias in aliases),
		re.IGNORECASE,
	)
	return [
		str(block["id"])
		for block in blocks
		if pattern.search(str(block["text"]))
	]

File: apps/worker-new/src/domain/normalizer.ts
import { loadVocabulary } from "./vocabulary.ts";

export interface ResumeBlock { readonly id: string; readonly text: string; readonly [key: string]: unknown }

export function normalizeResume(blocks: Iterable<ResumeBlock>): Record<string, unknown> {
	const blockList = [...blocks];
	const vocabulary = loadVocabulary();
	const skillHits = new Map<string, string[]>();
	for (const block of blockList) for (const name of Object.keys(vocabulary.mention(block.text))) {
		const ids = skillHits.get(name) ?? [];
		if (!ids.includes(block.id)) ids.push(block.id);
		skillHits.set(name, ids);
	}
	const order = new Map(blockList.map((block, index) => [block.id, index]));
	return {
		contact: contactFacts(blockList),
		skills: [...skillHits.keys()].sort((a, b) => a.localeCompare(b, undefined, { sensitivity: "base" })).map((canonicalName) => ({ canonicalName, category: vocabulary.categoryFor(canonicalName), evidenceBlockIds: [...(skillHits.get(canonicalName) ?? [])].sort((a, b) => (order.get(a) ?? blockList.length) - (order.get(b) ?? blockList.length)) })),
	};
}

export function contactFacts(blocks: Iterable<ResumeBlock>): { name: string | null; email: string | null; location: string | null } {
	const lines = [...blocks].flatMap((block) => block.text.split(/\r?\n/).map((line) => line.trim()).filter(Boolean));
	const combined = lines.join("\n");
	const email = combined.match(/\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/i)?.[0] ?? null;
	const first = lines[0] ?? "";
	const words = first.split(" ");
	const name = words.length >= 2 && words.length <= 5 && /^[A-Za-z][A-Za-z .'-]{1,80}$/.test(first) && words.every((word) => /^[A-Z]/.test(word)) ? first : null;
	const location = lines.slice(0, 12).find((line) => line.toLowerCase().startsWith("location:"))?.slice("Location:".length).trim() ?? null;
	return { name, email, location };
}
